### ISSUE TITLE
Use retryable HTTP code when HTTP API call fails with DeadlineExceeded = context deadline exceeded

### ISSUE THREAD

**Issue by pre** [argoproj/argo-cd#13038]
Checklist:

* [X] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [X] I've included steps to reproduce the bug.
* [X] I've pasted the output of `argocd version`.

**Describe the bug**

We access the ArgoCD Application HTTP API programmatically. Every now and then ArgoCD has too many Applications reconciling or updating at once, which results in HTTP 400 response with `code = DeadlineExceeded desc = context deadline exceeded` body. 

The HTTP Response Code is currently HTTP 400 "Bad Request" which signals that the error is permanent and should not be retried. However, the Context Deadline Exceeded is definitely retryable. After the ArgoCD load has passed, the exactly same HTTP API call to create an ArgoCD Application will succeed.

We can for sure read the response body for `code = DeadlineExceeded` to distinguish a retryable DeadlineExceeded timeout error from a non-retryable "Bad Request" kind of error. 

However, a more appropriate HTTP error code, such as HTTP 503 Resource Temporarily Unavailable could be more proper in the case of `code = DeadlineExceeded desc = context deadline exceeded` instead of HTTP 400 Bad Request.

**To Reproduce**

Have enough load in ArgoCD and try to create an Application resource using the ArgoCD HTTP API. 

**Expected behavior**

A HTTP response code which signals that the error is intermittent and request can be retried. 

**Screenshots**

<img width="1221" alt="image" src="https://user-images.githubusercontent.com/63895/228466131-0efeff1f-dc52-4d62-8098-9a10784b7df1.png">

<img width="471" alt="image" src="https://user-images.githubusercontent.com/63895/228466285-e03eda21-6605-4648-8e61-dcabb722d7a6.png">

**Version**

```shell
2.3.12
```

**Logs**

```
rpc error: code = DeadlineExceeded desc = context deadline exceeded
```

**Comment by todaywasawesome** [comment]
This isn't a bug so I've added an enhancement request to it. I'm not sure 503 would make sense because the 404 is generated by the client when it can't find the resource. The client has no idea if the context deadline has been exceeded because something is permanently not working or if it's just intermittent. 

The issue you're having can probably be resolved by scaling repo-server. I suggest reviewing this operator manual on scaling https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/

**Comment by pre** [comment]
>  I'm not sure 503 would make sense because the 404 is generated by the client when it can't find the resource

Note: The response is not HTTP 404 but HTTP 400. This not about being able to find a resource - this error is about ArgoCD not being able to *fulfill* the request under load.

The nature of the ArgoCD error `code = DeadlineExceeded`  is a timeout. The recovery step for the HTTP client is to retry the ArgoCD HTTP API request later, and the request will succeed once ArgoCD is no longer under load.

The nature of HTTP 400 is Bad Request which signals that the error was more likely caused by the HTTP client. However, this is not the case, as the error is caused by ArgoCD having too much load. In other words, it's the ArgoCD not being able to fulfill the request.

To be clear - the `code = DeadlineExceeded` appears in the HTTP 400 response's body and is generated by the argocd-server (argocd component which serves the http api). This is no way a client error and HTTP 400 is not the correct response code.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400

> The HyperText Transfer Protocol (HTTP) 400 Bad Request response status code indicates that the server cannot or will not process the request due to something that is perceived to be a client error (for example, malformed request syntax, invalid request message framing, or deceptive request routing).
>
> Warning: The client should not repeat this request without modification.

**Comment by pre** [comment]
FWIW we're read the argocd operator manual on scaling and applied all the findings. The problem is every now and then there's too much happening at once, which causes the timeout. 

We're mitigating by our own throttle mechanism on our client which calls the ArgoCD HTTP API. But it'd be much cleaner to retry on based on the HTTP error code instead of having to inspect the response body for `code = DeadlineExceeded` to know whether the error is retryable. By default we don't retry a HTTP 400 response.

**Comment by todaywasawesome** [comment]
@pre Makes sense, I'll bring it up in today's contrib experience meeting. Thanks for walking through it 😄

**Comment by leoluz** [comment]
Discussing in the contrib meeting we understand that this is probably an error handling issue that needs to be addressed in the code

**Comment by crenshaw-dev** [comment]
I believe the error from the first screenshot comes from here: https://github.com/argoproj/argo-cd/blob/9f6e5f93c0b310462ca5e49c80a18c5ef0b2ae1d/server/application/application.go#L1117

When we handle the error from `argo.ValidateRepo`, we should differentiate between a timeout and other errors.
